Add cfd_protocol crate to workspace

luckysori opened the pull request:
We've added a test which should prove that the spend transactions we generate in the protocol are valid. Some further assertions may be missing at this stage.

Currently it is all test code, so defining useful library APIs will come next. That will also include off-chain renewal of the CFD.

cc @da-kami @klochowicz

luckysori committed:
Generate all DLC transactions in test

luckysori committed:
Sign and encsign all DLC transactions in test

luckysori committed:
Remove erroneous note

luckysori committed:
Verify refund signatures and CET encsigs

The protocol doesn't need to (can?) verify the lock transaction signatures
and the broadcasting party should instead attempt to broadcast the
lock transaction.

luckysori committed:
Verify refund signatures using bitcoinconsensus

Co-authored-by: Philipp Hoenisch <philipp@hoenisch.at>

luckysori committed:
Verify signed CETs using bitcoinconsensus

luckysori committed:
Create cargo workspace and cfd_protocol crate

luckysori committed:
Commit Cargo.lock

luckysori committed:
Define possible library APIs

luckysori committed:
Introduce revocation-style transactions ala Thor

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -1,31 +1,154 @@
+use secp256k1_zkp::{Secp256k1, SecretKey};
+use std::collections::HashMap;
+
+// TODO: Consider extension trait
```
+1

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -1,31 +1,154 @@
+    pub dlc_pk: PublicKey,
+    pub dlc_amount: Amount,
+    pub address: Address,
+}
```
Creating one of these requires funding the psbt correct?

Maybe a extension trait above should already return you a `Party` struct?

thomaseizinger replied:
The wallet could fund the psbt and automatically fill in the address!

luckysori replied:
Great idea! The only thing that's less obvious is who should generate the corresponding `dlc_sk`? I'm leaning towards generating before and passing in `dlc_pk` to the trait method.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -646,11 +648,110 @@ mod tests {
+            let descriptor_str = format!(
+                "wsh(c:or_i(or_i({},{}),{}))",
+                maker_punish_condition, taker_punish_condition, cet_or_refund_condition
+            );
```
Have you considered using the miniscript compiler to create a miniscript from a policy instead of handrolling it? Might yield a more optimal version!

luckysori replied:
Good point! We pretty much copied what we already had in `thor`, but we did leave a TODO about optimizing the miniscript. We'll try to use the miniscript compiler once we tackle that TODO.

luckysori replied:
I've created https://github.com/comit-network/hermes/issues/19 to track this.

bonomat replied:
I used the online compiler for the updated script.

luckysori replied:
Ah, I didn't know that! It would still be good to have the code generate the miniscript from the policy, so I think we can leave the issue open.

luckysori committed:
Remove old FIXME

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -558,31 +621,161 @@ mod tests {
+            revoked_commit_tx: Transaction,
+        ) -> Result<Self> {
+            // CommitTransaction has only one input
+            let input = revoked_commit_tx.input[0].clone();
```
With `Itertools`, you could encode this invariant using `into_iter` and `exactly_one`: https://docs.rs/itertools/0.10.1/itertools/trait.Itertools.html#method.exactly_one

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -558,31 +621,161 @@ mod tests {
+                        continue;
+                    }
+                }
+            }
```
Such kind of stuff can be nicely extracted into extension traits so the actual flow of this function is less noisy and composed of more powerful low-level functionality.

luckysori replied:
Interesting idea. I rarely consider using the extension trait pattern, so it's very useful to get feedback on where they can be used!

In this case, would your idea be to extend TxIn with (e.g.) TxInExt and define a extract_signatures trait method? I guess in this particular instance since this all remains private API of the library we could just extract a free function too.

thomaseizinger replied:
Yeah, the only advantage of an extension trait over a free function is that it is easier to discover because the IDE will give you auto-completion for it.

In this case, yes probably put it on `TxIn`.

luckysori replied:
We haven't done this yet because it's just refactoring, but I would like to try this out: https://github.com/comit-network/hermes/issues/20. 

I could have probably just done it, but we can use that issue to remember that there might be other opportunities to use extension traits.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -558,31 +621,161 @@ mod tests {
+                let mut satisfier = HashMap::with_capacity(3);
+
+                let pk = bitcoin::secp256k1::PublicKey::from_secret_key(SECP256K1, &sk);
+                let pk_hash = hash160::Hash::hash(&pk.serialize()[..]);
```
https://docs.rs/bitcoin/0.27.0/bitcoin/util/ecdsa/struct.PublicKey.html#method.pubkey_hash

thomaseizinger reviewed (commented):
Some thoughts, feel free to ignore for now :)

luckysori committed:
Verify punish transactions

luckysori committed:
Refine possible public API and implement build_cfd_transactions

The next step should be to actually use `build_cfd_transactions` in
the test.

Afterwards, we can continue implementing and refining the public API of
the library.

bonomat committed:
Calculate CET fee based on tx size.

bonomat committed:
Use pubkey_hash function on PK

bonomat committed:
Use itertools exactly_one

luckysori committed:
Rewrite protocol test to use some public APIs

bonomat committed:
Simplify constructor by only passing in a sk

bonomat committed:
Implement build_party_params for Wallet

luckysori committed:
Use all public APIs in protocol test

thomaseizinger committed:
Avoid cloning and allocation of tx inputs

If all we need is a mutable reference to the tx input, we can
grab one via `iter_mut`.

bonomat commented on cfd_protocol/src/lib.rs:
```diff
@@ -246,7 +242,7 @@ pub fn punish_transaction(
     revoked_commit_tx: &Transaction,
 ) -> Result<Transaction> {
     // CommitTransaction has only one input
-    let input = revoked_commit_tx.input.clone().into_iter().exactly_one()?;
```
Unfortunately this breaks the build with: 

```
error[E0621]: explicit lifetime required in the type of `revoked_commit_tx`
   --> cfd_protocol/src/lib.rs:246:61
    |
243 |     revoked_commit_tx: &Transaction,
    |                        ------------ help: add explicit lifetime `'static` to the type of `revoked_commit_tx`: `&'static bitcoin::Transaction`
...
246 |     let input = revoked_commit_tx.input.iter().exactly_one()?;
    |                                                             ^ lifetime `'static` required

```

bonomat committed:
Optimize miniscript by combining the full expression

bonomat committed:
Subtract minimal fee from punish tx

bonomat committed:
Subtract minimal fee from commit tx

bonomat committed:
Calculate refund fee dynamically

luckysori committed:
Simplify extraction of publish sk from candidate commit tx

luckysori committed:
Remove unwrap

luckysori committed:
Remove comment and use anyhow::context instead

luckysori committed:
Add constructor for Payout

bonomat committed:
Add and run dprint fmt

bonomat committed:
Add ci

Includes a few checks:
- dprint check
- clippy
- workspace tests on ubuntu and macos

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -162,13 +520,375 @@ mod tests {
+    ) -> Result<()> {
+        encsig.verify(
+            SECP256K1,
+            &secp256k1_zkp::Message::from_slice(&self.sighash).expect("sighash is valid message"),
```
I think all the hash types of the `rust-bitcoin` ecosystem have fail-safe `From` implementations via this `ThirtyTwoByteHash` marker trait?

https://docs.rs/secp256k1/0.20.2/src/secp256k1/lib.rs.html#504-509

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -162,13 +520,375 @@ mod tests {
+        let lock_output = TxOut {
+            value: amount.as_sat(),
+            script_pubkey: lock_descriptor
+                .address(Network::Regtest)
```
Don't hardcode any networks please :)

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -22,6 +22,9 @@ use std::collections::HashMap;
 const MIN_RELAY_FEE: u64 = 1;
 
+// TODO use Amount type
+const P2PKH_DUST_LIMIT: u64 = 546;
```
I thin the recent rust-bitcoin version has a constant for this!

luckysori replied:
I think Philipp tackled this in a1cc6badb21de442403f4f8fb046b139076d498e.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -543,6 +544,14 @@ impl PunishTransaction {
+                SECP256K1,
+                &PrivateKey {
+                    compressed: true,
+                    network: Network::Regtest,
```
Same here, no hard-coded network please!

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -13,6 +13,7 @@ use bdk::{
     descriptor::Descriptor,
     miniscript::{descriptor::Wsh, DescriptorTrait},
 };
+use bitcoin::PrivateKey;
```
Not sure if better but maybe the protocol should work with `secp256k1::SecretKey`s instead of the `bitcoin` ones?

luckysori replied:
Good catch. This just came in because we copied over something from `thor` (which didn't need this type anyway :stuck_out_tongue:).

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -28,12 +30,27 @@ const MIN_RELAY_FEE: u64 = 1;
+        let mut builder = self.build_tx();
+        builder
+            .ordering(bdk::wallet::tx_builder::TxOrdering::Bip69Lexicographic)
+            .add_recipient(Script::new(), amount.as_sat());
```
Can we have a comment here why it is okay to pass an empty script? Does the size of the script here affect fee calculation?

bonomat replied:
This could be a source of errors as we are underpaying for the fees. 
@luckysori : I think we didn't notice this because I added  `.fee_rate(FeeRate::from_sat_per_vb(2.0))`. 

Did we run the docker-test with `1.0 sat/vbyte` as well?

bonomat replied:
Can confirm, this is an issue if fees are set to 1.0
` RpcError { code: -26, message: "min relay fee not met, 238 < 252", data: None }`

bonomat replied:
Fixed.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -209,12 +208,157 @@ pub fn spending_tx_sighash(
+    (maker_pk, maker_sig): (PublicKey, Signature),
+    (taker_pk, taker_sig): (PublicKey, Signature),
+) -> Result<Transaction> {
+    let satisfier = {
+        let mut satisfier = HashMap::with_capacity(2);
+
+        satisfier.insert(maker_pk, (maker_sig, SigHashType::All));
+        satisfier.insert(taker_pk, (taker_sig, SigHashType::All));
+
+        satisfier
+    };
```
Might be able to condense this using the `from_iter` function on `HashMap`: https://doc.rust-lang.org/std/collections/struct.HashMap.html#impl-FromIterator%3C(K%2C%20V)%3E

```rust
HashMap::from_iter(vec![
	(maker_pk, (maker_sig, SigHashType::All)),
	(taker_pk, (taker_sig, SigHashType::All))
]);
```

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -209,12 +208,157 @@ pub fn spending_tx_sighash(
+    let publish_them_sk = sigs
+        .into_iter()
+        .find_map(|sig| encsig.recover(SECP256K1, &sig, &publish_them_pk.key).ok())
+        .context("Could not recover secret key from revoked transaction")?;
```
Wild idea: To condense this down, could extract a custom `Iterator` extension trait that works on `Iterator<Item = Signature>`.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -255,7 +399,7 @@ pub struct Payout {
 
 impl Payout {
-    fn to_txouts(&self, maker_address: &Address, taker_address: &Address) -> Vec<TxOut> {
+    fn as_txouts(&self, maker_address: &Address, taker_address: &Address) -> Vec<TxOut> {
```
`as_` should be a cheap conversion by convention. You are returning an owned instance from a reference (`&self`) which implies an allocation. I think this should be named `to_`.

Did clippy suggest this?

luckysori replied:
I blame Clippy! (It was Clippy)

thomaseizinger reviewed (approved):
Nice work, thank you!

Should be a great foundation to build on top of!
It would be nice if we can gradually move towards extracting a test harness for Bitcoin transactions that combines consensus-based tests with all kinds of other checks that we learn over time are needed :)

luckysori committed:
Handle fees properly

bonomat committed:
Compute dust limit on demand

bonomat committed:
Use dummy multisig for creating lock psbt

bonomat committed:
Fix clippy warnings: unnecessary results
